mimxrt/machine pin: Added machine_pin module

alphaFred opened the pull request:
Implemented basic `Pin` class for `machine` module featuring the following: 
- implemented and integrated `make-pins.py` script
- Configuration of GPIO for input/output/open drain output
- Configuration of initial value for outputs
- `on`/`off`, `low`/`high`, `value`, `init`, `drive` methods
- added options to control drive strength, and pull settings

Additional fixes/optimizations for testing and general improvements: 
- Setup of CI for mimxrt port
- Addition of `LDFLAGS += --gc-sections` for DEBUG build to allow debugging
- make target for deployment via JLINK to allow easier development

Implements #7164

alphaFred committed:
mimxrt/machine_pin: Started rework of machine_pin.

- modified pin type from pin_obj_t to machine_pin_obj_t
- created machine_pin.c
- created pin.csv and mimxrt1021_af.csv files
- started integration of new API into pin.c/.h

alphaFred committed:
mimxrt/machine_pin: Implemented basic make-pins.py.

- added all alternative functions for pins of  mimxrt1021-evk board
which are available through arduino pin header
- added arduino and pad names of mimxrt1021-evk board in respective
pins.csv
- implemented basic version of make-pins.py to genertate pins.c/.h files
automatically. The only alternate function currently supported is GPIO.

alphaFred committed:
mimxrt/machine_pin: Implemented basic machine_pin.c.

- updated make-pins to generate *pin_obj array and
  *pins_locals_dicts correctly
- created machine_pin.c file for implementation of pin module

alphaFred committed:
mimxrt/machine_pin: Implemented compilable machine_pin.

Implemented bare minimum compilable version of machine_pin.c
with automatic generation of pins through .csv file.

alphaFred committed:
mimxrt/machine_pin: Modified DEBUG LDFLAGS.

Modified LDFLAGS for DEBUG build to get usefull .elf file for
debugging with OZONE.

alphaFred committed:
mimxrt/machine_pin: Added required mp_hal callback.

Added required mp_hal callback mp_hal_time_ns.

alphaFred committed:
mimxrt/machine_pin: Implemented on/off methods.

Implemented on and off method for Pin class.

alphaFred committed:
mimxrt/machine_pin: Implemented IN/OUT mode.

Implemented mode argument with IN and OUT as options.
Additionally reworked pin_find_af function and
machine_pin_af structure

alphaFred committed:
tools/ci: Adding CI for mimxrt port.

Added github workflow and ci functions for mimxrt port.

alphaFred committed:
mimxrt: Enabled built-in help.

alphaFred committed:
mimxrt: Fixed JLINK deploy command.

alphaFred committed:
mimxrt/machine_pin: Implemented value argument.

alphaFred committed:
mimxrt/machine_pin: Added general improvements.

alphaFred committed:
mimxrt/machine_pin: Implemented value method.

alphaFred committed:
mimxrt/machine_pin: Implemented mode method.

Implemented mode method and upgraded make-pins.py and
related files to not treat machine_pin objects as constanst.
Unfortunately reading out the current configuration of a PAD is
relatively complex. There is no function provided by the NXP HAL
afaik and therefore it is easier to implement a field in the
machine_pin object which contains that information. The downside
of that is increased RAM usage.

alphaFred committed:
Merge branch 'upstream_master' into mimxrt/machine_pin

alphaFred committed:
mimxrt/machine_pin: Updated all boards with .csv.

Updated all board folders with .csv files for automatic pin generation.
Updated README.md.

Signed-off-by: Philipp Ebensberger

alphaFred committed:
Merge branch 'upstream_master' into mimxrt/machine_pin

alphaFred committed:
mimxrt/machine_pin: Fixed pad configuration for input.

alphaFred committed:
mimxrt/machine_pin: Cleaned up make-pins.py.

alphaFred commented:
I am confident about the change now. Tested the new features on my MIMXRT-1020 EVK. Therefore I humbly ask for a review if you find the time 😊. 
@dpgeorge @jimmo

robert-hh commented:
Thank you for reviving the work for mimxrt. It's a good chip and deserves better support. And the Pin class you made is surely a big chunk of work and a great start. I could build it for Teensy4.0. and define pins, setting them to IN and OUT.
What do I miss comparing it  to other ports:

- setting the pin level with pin.value(arg); that is often used in programs since especially on() and off() is misleading.
- setting the pin level with pin(arg); fast track for pin.value()
- defining the pull modes for input; You mentioned that already in the code.

alphaFred commented:
Hi @robert-hh thanks a lot for your support. I totally agree with your assessment. As a matter of fact I already started working on these topics. However I believe steady improvement is better then Big Bang commits. 

Therefore I would love to get this first step merged and then follow up. The i.MX certainly has a lot of options for configuring pins ☺️.

alphaFred committed:
mimxrt/machine_pin: Implemented setting pin via value.

alphaFred committed:
mimxrt/machine_pin: Implemented call class method.

alphaFred committed:
mimxrt/machine_pin: Implemented configuration of pull settings.

robert-hh reviewed (commented):
That's interesting. Do you know whether the keeper and pull options are exclusive or can be combined. Especially keeper seems to be more related to an output state than an input state. But that's just the first impression of a short glance into the manuals.
Edit: I've & tested already the pin.value() and pin() options.

alphaFred commented:
That is a good question. The manual is really poor on details here. In the reference manual many of the options are actually not discusses at all. The whole GPIO and Pin settings in this chip are very strange and cumbersome - at least to me.
In addition to the reference manual there is an application note available from NXP, providing some more details [AN5078 Influence of Pin Setting on System Function and Performance](https://www.nxp.com/docs/en/application-note/AN5078.pdf)

From what I understood the keeper circuit is essentially a latch which "holds" the last input/output level. It has a high impedance and basically acts as an pull-up or pull-down depending on the last level. I found an application not from ST ([AN2149 Integrated bus hold circuitry](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwj7m9KfkLXwAhUHDxQKHQNiAdcQFjABegQICRAD&url=https%3A%2F%2Fwww.st.com%2Fresource%2Fen%2Fapplication_note%2Fcd00055065-integrated-bus-hold-circuitry-stmicroelectronics.pdf&usg=AOvVaw1nGgKwYdJlSHe8j6MzZRPG)) which has a bit more details on bus-keeper circuits and in this document they also specifically discuss inputs. So I assume there is a good use case in either way.

robert-hh commented:
But keeping the last state when? For the other aspects I can confirm that the settings for the PULL_UP variants are assigned properly with ~100k, ~47k and ~22K, and that PULL_DOWN is indeed ~100k.

alphaFred commented:
When your input/output goes high-Z. In the NXP application it is mentioned that the keeper will hold the last output value when you enable the output driver. Which is essentially disconnecting the pin - e.g. high impedance. I have read a similar example where your peripheral is disconnected. Instead of having a fix "default" value which you would get from a pull-up/down resistor the input/output will retain the last state (high or low). At least that is my understanding so take it with a grain of salt.

robert-hh commented:
I found also this paper: https://www.nxp.com/docs/en/application-note/AN5078.pdf
I read it that the keeper plays a role when the output driver is disabled. But that option is not supported yet in the API.

alphaFred commented:
Yes that is correct. I have not implemented the drive strength options, yet. When you configure an input I choose to disable the output driver automatically. When you configure an output the drive strength will be configured as R0/4 which means 37Ohms @ 3.3V (this is the intermediate setting).

robert-hh commented:
The typical implementation in MicroPython is, that the init() call and the make_new() call use the same underlying init_helper() function, which sets up all parameters. For that reason, object instantiation and init() have the same set of options. As a sometimes surprising side effect you cannot change just one parameter with init() for most devices.
Just compare that e.g. to the esp32 or rp2 port. Similar, pin.value() and p() use the same function, which you duplicated in the code.

alphaFred commented:
What I meant is that if the `init` method is also implemented you could disable the output driver of an already existing pin instance. Creating the scenario described in the application note. Of course this option would also be available to the `make_new` function and you could directly instant iahte a pin with disabled output driver.

robert-hh commented:
It might be an option to take away the PULL_KEEPER from the pull modes (because it is not a pull mode) and define a separate method for output drive strength, with options like STRONG, KEEP, ....

alphaFred commented:
I beg to differ on that one. Keeper or pull settings have nothing to do with drive strength. Hence I would argue that the keeper setting should stay with the `pull` parameter as it is now. Actually you really would like to configure both options independently to cater to your circuit's needs.

robert-hh commented:
I was not thinking in terms of the controller design, but more in terms of usage. the keeper options does not seem to belong to an behavior as input, like the pull options do. As far as I understand, it defines more a behavior as output, and then it is a very high impedance drive strength; not 37Ohm, but ~180kOhm.

alphaFred commented:
The keeper has the same purpose as a pull-up or -down. Keeping the pin from floating. This is why I would argue the current design is acceptable in that regard.

alphaFred committed:
mimxrt/machine_pin: Implemented init method.

alphaFred committed:
mimxrt/machine_pin: Renamed PULL_KEEPER.

Renamed PULL_KEEPER to PULL_HOLD as described in the official
documentation.

robert-hh commented:
I made a few tests on the actual behavior of a pin with keeper enabled, and it actually latches the last stable input value. Interesting. So you are right. Must be some kind of positive fed back logic. I wonder what it's good for.

From reading the AN5078 doc I had the impression that it latches the state of the output register. But it does not.

alphaFred commented:
Check out the respective Wikipedia page. [Bus-holder](https://en.wikipedia.org/wiki/Bus-holder). On a bus this makes a lot of sense. 

I will implement drive strength, speed, slew rate and hysteresis settings in the coming days. I guess the PR will be open for a couple more days 🙈.

May I ask which board you are using for testing?

robert-hh commented:
Don't do more than provided by the other ports. The first step should be to make it compatible, both for the API set and internal design. So drive seems OK. Setting  slew rate and hysteresis does not exist on other ports. OPEN_DRAIN mode is important. So I recommend to add that next. It is needed for some popular sensors, and once you had the had on it, It is easy to add SoftI2C. 
And yes. There is still time to work on the PR.
The more challenging part for implementation would be interrupt handling.
I'm using a Teensy 4.0 with the MIMXRT1062DVL6A chip.

robert-hh commented:
Hi @alphaFred How is work going on? 
Meanwhile I started the attempt to add the flash file system package. This goes well besides the usual seemingly simple things, that can hold one up for hours. As I first step I have embedding frozen bytecode working, and the Skeleton of the block device driver for flash. It supports already block read and some IOCTL commands. What is missing now is the code for flash write and flash erase. There seems no official library support by NXP for the hardware - or I did not find it yet. What I found were a few implementations, like these from Paul Stoffregen @ PJRC for teensyduino, and by Rocky Song for the OpenMV platform at https://github.com/RockySong/micropython-rocky. The latter could be a good source of hints for implementations, although the repository itself is based on a older version of MP. Especially since if uses the MIT License. What Rocky uses are calls to flexspi_nor_flash_page_program() and flexspi_nor_flash_erase_sector(), which seem to be the lib calls that I need, but I did not locate yet where these are located.
The actual state of the game is at https://github.com/robert-hh/micropython/tree/mimxrt, which includes the recent state of your commit. If we both work on different parts of the implementation, we have to find a proper way of collaboration.

alphaFred commented:
Hi @robert-hh sounds really great. The flash file system would have been my next step as well ☺️. What I could see from Rocky Song's implementation is that he has written these wrappers (`flexspi_nor_flash_page_program()`,`flexspi_nor_flash_erase_sector()`) for him self in `flash.h` . NXP HAL does only provide an interface to the flex SPI peripheral as far as I know. 

I am quite busy over the weekend meaning I will probably not find the time. Hence I will continue on Monday.

alphaFred committed:
mimxrt/machine_pin: Added init ad drive methods.

Added option to configure the drive strength of pins
(used for impedance matching). Since i.MXRT supports more options
there than other ports in addition to the known LOW, MED, and HIGH
intermediate levels are available. On top of that the output driver
can even be disabled for inputs for instance.

alphaFred committed:
mimxrt/machine_pin: Fixed review findings.

robert-hh commented:
Thanks @alphaFred . That is really improving. I made a few first tests, and the new additions work, almost as expected. I personally do not like the many options for the drive keyword, and the length of each option word. Even if the controller supports 7 + 1 drive modes, maybe 3 + 1 are sufficient for exposition by the interface, like HIGH, MED, LOW, OFF.

alphaFred commented:
Hi @dpgeorge or @jimmo. If any of you finds the time it would be great if you could review the PR please. Many thanks in advance.

alphaFred committed:
mimxrt/machine_pin: Implemented open drain mode.

dpgeorge commented on ports/mimxrt/boards/MIMXRT1010_EVK/MIMXRT1010_EVK_af.csv:
```diff
@@ -0,0 +1,22 @@
+Pad,ALT0, ALT1, ALT2, ALT3, ALT4, ALT5, ALT6, ALT7, ALT8, ALT9,ADC,ACMP,Default
```
I think this file should go up one directory, in `ports/mimxrt/boards/`, because it's useful for all RT1010-based boards, not just this EVK.

alphaFred replied:
I do not think this is a good idea, frankly because the content of these files is specific to each board. There are overlaps but that's it. I actually only added the pins that are available through the Arduino Header (for the *_EVK boards). Therefore it is not useful in my perspective as a "generic" documents for a chip variant. At least not for now.

dpgeorge commented on ports/mimxrt/boards/MIMXRT1010_EVK/pins.csv:
```diff
@@ -0,0 +1,57 @@
+ENC_A,GPIO_AD_05
+ENC_B,GPIO_AD_06
+LED_GREEN,GPIO_11
+GPIO_01,GPIO_01
```
I think the `make-pins.py` script supports the notation that you can leave the first name out if it's the same as the second, eg `,GPIO_01`.

alphaFred replied:
Ok, I guess I have to modify my script for that. I have basically written it from scratch to get a better grasp.

dpgeorge replied:
Ah, I see.  Well, it should be easy to add this feature.  And I think it's quite useful to indicate which pins don't have a special name (just the datasheet name) but are still exposed to the user.

alphaFred replied:
Yes. I will fix that. 👍

dpgeorge commented on ports/mimxrt/boards/MIMXRT1020_EVK/MIMXRT1020_EVK_af.csv:
```diff
@@ -0,0 +1,23 @@
+Pad,ALT0, ALT1, ALT2, ALT3, ALT4, ALT5, ALT6, ALT7, ALT8, ALT9,ADC,ACMP,Default
+GPIO_AD_B1_09,LPI2C2_SDA,FLEXPWM1_PWMB1,LPUART2_RXD,SAI1_TX_DATA3,FLEXIO1_D06,GPIO1_IO25,LPSPI3_PCS1,XBAR_INOUT13,,,ADC1_IN9,ACMP2_IN5,ALT5
```
as above, this can be moved up a directory

alphaFred replied:
see above

dpgeorge commented on ports/mimxrt/boards/MIMXRT1050_EVK/MIMXRT1050_EVK_af.csv:
```diff
@@ -0,0 +1,21 @@
+Pad,ALT0, ALT1, ALT2, ALT3, ALT4, ALT5, ALT6, ALT7, ALT8, ALT9,ADC,ACMP,Default
+GPIO_AD_B1_07,FLEXSPI_B_DATA0,LPI2C3_SCL,LPUART3_RXD,SPDIF_EXT_CLK,CSI_HSYNC,GPIO1_IO23,USDHC2_DATA3,KPP_COL4,,,ADC1_IN12,ACMP1_IN5,ALT5
```
as above, this file can be moved up a directory

alphaFred replied:
see above

dpgeorge commented on ports/mimxrt/boards/MIMXRT1060_EVK/MIMXRT1060_EVK_af.csv:
```diff
@@ -0,0 +1,21 @@
+Pad,ALT0, ALT1, ALT2, ALT3, ALT4, ALT5, ALT6, ALT7, ALT8, ALT9,ADC,ACMP,Default
+GPIO_AD_B1_07,FLEXSPI_B_DATA0,LPI2C3_SCL,LPUART3_RXD,SPDIF_EXT_CLK,CSI_HSYNC,GPIO1_IO23,USDHC2_DATA3,KPP_COL4,GPT2_COMPARE3,FLEXIO3_D07,ADC1_IN12,ACMP1_IN5,ALT5
```
as above, this can be moved up a directory

alphaFred replied:
see above

dpgeorge commented on ports/mimxrt/boards/MIMXRT1064_EVK/MIMXRT1064_EVK_af.csv:
```diff
@@ -0,0 +1,21 @@
+Pad,ALT0, ALT1, ALT2, ALT3, ALT4, ALT5, ALT6, ALT7, ALT8, ALT9,ADC,ACMP,Default
+GPIO_AD_B1_07,FLEXSPI_B_DATA0,LPI2C3_SCL,LPUART3_RXD,SPDIF_EXT_CLK,CSI_HSYNC,GPIO1_IO23,USDHC2_DATA3,KPP_COL4,GPT2_COMPARE3,FLEXIO3_D07,ADC1_IN12,ACMP1_IN5,ALT5
```
as above, this can be moved up a directory

alphaFred replied:
see above

dpgeorge commented on ports/mimxrt/boards/TEENSY40/TEENSY40_af.csv:
```diff
@@ -0,0 +1,39 @@
+Pad,ALT0, ALT1, ALT2, ALT3, ALT4, ALT5, ALT6, ALT7, ALT8, ALT9,ADC,ACMP,Default
+GPIO_AD_B0_03,FLEXCAN2_RX,XBAR_INOUT17,LPUART6_RXD,USB_OTG1_OC,FLEXPWM1_PWM1_X,GPIO1_IO03,REF_CLK_24M,LPSPI3_PCS0,,,,ACMP4_IN4,ALT5
```
can this board use one of the existing af files?  why does it need to be specific to the Teensy 4.0?

alphaFred replied:
Teensy uses different pins than the other boards and has more pins available to the user. So I think there needs to be a separate file.

dpgeorge replied:
On stm32 there is one AF file per MCU and then different boards using the same MCU use the same AF file.  It doesn't matter if the board does not use all the pins, it just selects those pins from the AF file that it uses.

In other words: if board A and board B have the same i.MX MCU and expose the same pin (eg GPIO_AD_B0_03) does that pin have the exact same functions on both boards?  If so then they can share the same AF file.

alphaFred replied:
Mhmh yeah now I got your point. You are right. It makes sense to do it like that. I will fix that.

dpgeorge replied:
Great!

dpgeorge commented:
I didn't get time for a full review, but left some comments.

Please do keep this "minimal", and only add what is already documented for `machine.Pin`.  That makes it much less controversial and we only need to discuss implementation.

Regarding pin naming: is something like `GPIO_AD_B1_14` really a good name?  Then it'll be `machine.Pin("GPIO_AD_B1_14")` which seems rather long to remember and type... it could drop the `GPIO_` prefix.  Can pins just be numbered via integers, does that make sense, or are there also ports?

alphaFred commented:
I totally get your point here. Personally I also dislike the names but these are the actual pin (or rather pad to keep the NXP lango) names of the device. And unfortunately you can not easily (at least to my knowledge) number the pins.
There are different kind of pins which share the same "number". For instance:
You have `GPIO_AD_B0_00` which is `GPIO1_IO00` and `GPIO_EMC_00` which is `GPIO2_IO00` but you also have `GIPO_SD_B0_00` which is `GPIO3_IO13` (example from MIMXRT1021). Therefore any number etc. would only make it worse in my perspective.
There are the Arduino pin names like `D0` however.

robert-hh commented:
With Teensy I used Pin(14, ..) or Pin(13, ....) and it were the right ones.

alphaFred commented:
Yes that is correct, because teensy has the pins labeled with numbers in its documentation. The pin configuration .cvs files provide the option to give an arbitrary name to an actual pin. Therefore on teensy selecting pin `14` is linked with `GPIO_AD_B1_02`. 

But this particular identifier is only applicable for a particular board. 

The i.MX does not have the notion of numbered pins like for instance the RP2040 of the raspberry pi pico where GPIO are numbered from 0-29. Hence I can not have a single configuration for all boards. 

At least this is my understanding. I would be happy if I can simplify the whole thing.

alphaFred committed:
Merge branch 'upstream_master' into mimxrt/machine_pin

alphaFred committed:
tools/ci: Reduced number of boards to build.

Based on comments from @dpgeorge only MIMXRT1020EVK and TEENSY40
board builds are kept in the CI configuration to keep build
time low.

alphaFred committed:
mimxrt/machine_pin: Reworked af files.

- Created chip specific af files
- Reworked pin type for ROM storage
    - removed strage of mode in pin type
    - added functions to read out the mode from registers
- Renamed pins to inclusive names
    - MOSI/MISO -> SDO/SDI

alphaFred commented:
@dpgeorge rework of the af files is done. We now have a single AF file for each MCU Variant. Since we have 5 boards with 5 different MCUs, there are still 5 af files ☺️.

dpgeorge commented on ports/mimxrt/boards/MIMXRT1020_EVK/mpconfigboard.mk:
```diff
@@ -2,6 +2,25 @@ MCU_SERIES = MIMXRT1021
+
+
+deploy_jlink: $(BUILD)/firmware.hex
+	$(Q)$(TOUCH) $(JLINK_COMMANDER_SCRIPT)
```
is this touch command needed if the script is anyway rebuild from scratch below?

alphaFred replied:
Most likely not. I will take a look and remove if I can.

alphaFred replied:
removed

dpgeorge commented on ports/mimxrt/machine_pin.c:
```diff
@@ -0,0 +1,325 @@
+const mp_obj_type_t machine_pin_cpu_pins_obj_type = {
+    { &mp_type_type },
+    .name = MP_QSTR_cpu,
+    .print = named_pin_obj_print,
```
this `.print` entry is not needed (I don't think it's even possible to get to this code because there are no instances of this type)

alphaFred replied:
removed

dpgeorge commented on ports/mimxrt/machine_pin.c:
```diff
@@ -0,0 +1,325 @@
+const mp_obj_type_t machine_pin_board_pins_obj_type = {
+    { &mp_type_type },
+    .name = MP_QSTR_board,
+    .print = named_pin_obj_print,
```
as above, this is not needed

alphaFred replied:
removed

dpgeorge commented on ports/mimxrt/machine_pin.c:
```diff
@@ -0,0 +1,325 @@
+            mp_hal_pin_high(self);
+        } else {
+            mp_hal_pin_low(self);
+        }
```
might be good to add an `mp_hal_pin_write(pin, value)` macro and then here it'd just be:
```c
mp_hal_pin_write(self, mp_obj_is_true(args[0]));
```

alphaFred replied:
Implemented and fixed

dpgeorge commented on ports/mimxrt/machine_pin.c:
```diff
@@ -0,0 +1,325 @@
+STATIC void machine_pin_obj_print(const mp_print_t *print, mp_obj_t o, mp_print_kind_t kind) {
+    (void)kind;
+    const machine_pin_obj_t *self = MP_OBJ_TO_PTR(o);
+    mp_printf(print, "PIN(%s)", qstr_str(self->name));
```
this should be `Pin(%s)`

alphaFred replied:
Fixed

dpgeorge commented on ports/mimxrt/machine_pin.c:
```diff
@@ -0,0 +1,325 @@
+    { MP_ROM_QSTR(MP_QSTR_value),   MP_ROM_PTR(&machine_pin_value_obj) },
+    { MP_ROM_QSTR(MP_QSTR_mode),    MP_ROM_PTR(&machine_pin_mode_obj) },
+    { MP_ROM_QSTR(MP_QSTR_init),    MP_ROM_PTR(&machine_pin_init_obj) },
+    { MP_ROM_QSTR(MP_QSTR_drive),   MP_ROM_PTR(&machine_pin_drive_obj) },
```
For now, please don't implement `mode()` or `drive()` methods.  They are not part of the core Pin API and the functionality of setting these values should be available via `init()`.

alphaFred replied:
Ok. 👌

alphaFred replied:
`drive()` and  `mode()`methods have been removed

dpgeorge commented:
Regarding pin naming: on Teensy 4.0 I get the following available pins (CPU name):
```
>>> machine.Pin.cpu.GPIO_
GPIO_AD_B0_00   GPIO_AD_B0_01   GPIO_AD_B0_02   GPIO_AD_B0_03
GPIO_AD_B0_04   GPIO_AD_B0_05   GPIO_AD_B0_06   GPIO_AD_B0_07
GPIO_AD_B0_08   GPIO_AD_B0_09   GPIO_AD_B0_10   GPIO_AD_B0_11
GPIO_AD_B0_12   GPIO_AD_B0_13   GPIO_AD_B0_14   GPIO_AD_B0_15
GPIO_AD_B1_00   GPIO_AD_B1_01   GPIO_AD_B1_02   GPIO_AD_B1_03
GPIO_AD_B1_04   GPIO_AD_B1_05   GPIO_AD_B1_06   GPIO_AD_B1_07
GPIO_AD_B1_08   GPIO_AD_B1_09   GPIO_AD_B1_10   GPIO_AD_B1_11
GPIO_AD_B1_12   GPIO_AD_B1_13   GPIO_AD_B1_14   GPIO_AD_B1_15
GPIO_B0_00      GPIO_B0_01      GPIO_B0_02      GPIO_B0_03
GPIO_B0_04      GPIO_B0_05      GPIO_B0_06      GPIO_B0_07
GPIO_B0_08      GPIO_B0_09      GPIO_B0_10      GPIO_B0_11
GPIO_B0_12      GPIO_B0_13      GPIO_B0_14      GPIO_B0_15
GPIO_B1_00      GPIO_B1_01      GPIO_B1_02      GPIO_B1_03
GPIO_B1_04      GPIO_B1_05      GPIO_B1_06      GPIO_B1_07
GPIO_B1_08      GPIO_B1_09      GPIO_B1_10      GPIO_B1_11
GPIO_B1_12      GPIO_B1_13      GPIO_B1_14      GPIO_B1_15
GPIO_EMC_00     GPIO_EMC_01     GPIO_EMC_02     GPIO_EMC_03
GPIO_EMC_04     GPIO_EMC_05     GPIO_EMC_06     GPIO_EMC_07
GPIO_EMC_08     GPIO_EMC_09     GPIO_EMC_10     GPIO_EMC_11
GPIO_EMC_12     GPIO_EMC_13     GPIO_EMC_14     GPIO_EMC_15
GPIO_EMC_16     GPIO_EMC_17     GPIO_EMC_18     GPIO_EMC_19
GPIO_EMC_20     GPIO_EMC_21     GPIO_EMC_22     GPIO_EMC_23
GPIO_EMC_24     GPIO_EMC_25     GPIO_EMC_26     GPIO_EMC_27
GPIO_EMC_28     GPIO_EMC_29     GPIO_EMC_30     GPIO_EMC_31
GPIO_EMC_32     GPIO_EMC_33     GPIO_EMC_34     GPIO_EMC_35
GPIO_EMC_36     GPIO_EMC_37     GPIO_EMC_38     GPIO_EMC_39
GPIO_EMC_40     GPIO_EMC_41     GPIO_SD_B0_00   GPIO_SD_B0_01
GPIO_SD_B0_02   GPIO_SD_B0_03   GPIO_SD_B0_04   GPIO_SD_B0_05
GPIO_SD_B1_00   GPIO_SD_B1_01   GPIO_SD_B1_02   GPIO_SD_B1_03
GPIO_SD_B1_04   GPIO_SD_B1_05   GPIO_SD_B1_06   GPIO_SD_B1_07
GPIO_SD_B1_08   GPIO_SD_B1_09   GPIO_SD_B1_10   GPIO_SD_B1_11
```
Although that's a lot of text, the names are a bit long, I don't think there's any sensible way to shorten them.  The names above correspond exactly to the names in the datasheet (of the "pad") and if they were shortened to something like "B1_00" that could be a bit confusing (in particular some pads are named "GPIO_01" and shortening that to "01" would be very confusing; and there's also "WAKEUP" but I don't know if that is a GPIO).

So I think the CPU names are OK as they are.  A given board should then make good use of board names/aliases to make it easier to construct a pin (and this is done in this PR).

-------------

Regarding naming of drive strength: IMO it's confusing naming them as LOWEST_POWER, LOW_POWER, LOWER_POWER, etc (is LOWER more or less than LOW?).  It could be enough to just have LOW_POWER, MED_POWER and HIGH_POWER.  Or if all are needed then maybe POWER0 through to POWER7?

alphaFred commented:
Yeah the drive strength names were a bit of a far stretch, I have to admit that. I tried to keep the existing naming as it is described in the documentation (LOW, MED, HIGH). I can see POWER0-POWER7 though. With POWER0 corresponding to lowest drive strength.

robert-hh commented:
Regarding the names for power strength the order is strange. I confirmed that by measuring. It does indeed not follow the language comparative.
LOWEST_POWER, LOW_POWER, LOWER_POWER, .. So LOW_POWER is lower than LOWER_POWER. So more an administrative ordering. When I added a 33 Ohm load to a Pin, I got the following voltages with the setting:
```
Highest - High - Higher - Med - Lower - Low - Lowest - Off

2.033 - 1.899 - 1.738 - 1,527 - 1.142 - 0.807 - 0.416 - 0 
```
P.S.: I rack my brain about this sequence of comparatives. I recall faintly a comedian sketch which recurred on that or something similar, but i cannot recall which one.

alphaFred commented:
Hi @robert-hh to be honest the measurement is exactly as I have envisioned it. I guess it is my fault but I meant `HIGHER` is above `MEDIUM` and lower than `HIGH`. Now that we spoke about it the order is also wrong in my eyes. I guess that whole confusion is testament to the bad naming :smile:

robert-hh commented:
No complaint at all. I made the test only for the verification of the "black box path" between the MicroPython code and the result at the output. This PR represent a huge amount of work, and is essential for all other I/O based modules for the MIMXRT family. And we all are happy that you kicked off the stall for these MCUs.

alphaFred commented:
No worries, I did not take it as a complaint. You are both right.

alphaFred committed:
mimxrt/machine_pin: Implemented review comments.

alphaFred commented:
on a side note, I could not sign off my commits because my name is too long. So unfortunately I could not check all the boxes in the CI build :smile::smile:

dpgeorge commented on ports/mimxrt/machine_pin.c:
```diff
@@ -294,13 +245,13 @@ STATIC const mp_rom_map_elem_t machine_pin_locals_dict_table[] = {
-    { MP_ROM_QSTR(MP_QSTR_HIGHER_POWER), MP_ROM_INT(PIN_DRIVE_HIGHER_POWER) },
-    { MP_ROM_QSTR(MP_QSTR_HIGH_POWER), MP_ROM_INT(PIN_DRIVE_HIGH_POWER) },
-    { MP_ROM_QSTR(MP_QSTR_HIGHEST_POWER), MP_ROM_INT(PIN_DRIVE_HIGHEST_POWER) },
+    { MP_ROM_QSTR(MP_QSTR_DRIVE_POWER_0), MP_ROM_INT(PIN_DRIVE_POWER_0) }, // R0 (150 Ohm @3.3V / 260 Ohm @ 1.8V)
```
For consistency, should `DRIVER_OFF` be renamed to `DRIVE_POWER_0`, and everything else here shifted up by 1?

Also, I always advocate for shorter names, and here I would consider `POWER0`, `POWER1`, etc.  In code that would look something like:
```python
p = Pin("GPIO_B_01", Pin.OUT, drive=Pin.POWER2)
```

alphaFred replied:
I really would like to keep the `DRIVE_OFF` option since this literally disables the output driver. With the rest I am fine and will fix it

alphaFred replied:
Fixed.

alphaFred committed:
mimxrt/machine_pin: Shortened names.

alphaFred committed:
mimxrt/machine_pin: Minor fixes in af files.

Thanks to @robert-hh a couple of missing entries could be fixed.
Further more we identified inconsistencies in the reference manual
and NXP driver libraries.

dpgeorge commented on ports/mimxrt/board_init.c:
```diff
@@ -39,7 +39,7 @@
 #include "clock_config.h"
 
-#define LED_STATE_ON (0)
+#define LED_STATE_ON (0)  // Todo: remove unused define
```
can this be removed now?

alphaFred replied:
yup - will be fixed

dpgeorge commented on ports/mimxrt/Makefile:
```diff
@@ -49,12 +59,14 @@ CFLAGS += -DXIP_EXTERNAL_FLASH=1 \
 ifeq ($(DEBUG),1)
 CFLAGS += -O0 -ggdb
+LDFLAGS += --gc-sections
+CFLAGS += -fdata-sections -ffunction-sections
```
why does this need to be added here?

alphaFred replied:
without it I could not see the symbols properly when debugging with JLink. It is "just" the DEBUG build so it has no influence on the release build

dpgeorge commented on ports/mimxrt/machine_pin.c:
```diff
@@ -0,0 +1,276 @@
+MP_DEFINE_CONST_FUN_OBJ_KW(machine_pin_init_obj, 1, machine_pin_init);
+
+STATIC const mp_rom_map_elem_t machine_pin_locals_dict_table[] = {
+    // TODO: Implement class locals dictionary
```
can this TODO comment be removed?

alphaFred replied:
yes - will be fixed

dpgeorge commented on ports/mimxrt/mphalport.h:
```diff
@@ -29,13 +29,22 @@
 
 void mp_hal_set_interrupt_char(int c);
 
+static inline uint64_t  mp_hal_time_ns(void) {
```
is this function needed for this PR to work?

alphaFred replied:
:thinking: I just tried compiling without it and it works. Interesting. I could swear that the build was failing without it - that is why I included it.
Ok - will be removed :+1:

alphaFred replied:
Yeah, could be. But we (myself and @robert-hh) made sure that both changes are completely stand-alone to make reviewing easier for you. Therefore there is no real need to keep it in this change.

dpgeorge commented:
I'm pretty happy with this now.  I've tested it on a Teensy 4.0 and it works.

There are many commits in this PR and they will need to be squashed before going in to master.  I will do this when rebasing and merging.  There will be one commit for adding mimxrt to the CI builds and one or two commits for the addition of machine.Pin (@alphaFred let me know if you have a problem with that).

alphaFred commented:
Hi @dpgeorge 
I will clean up your points in a minute. I have no problem when you squash the commits. I thought it would be better for me to keep track when I push most of my commits and do not perform heavy rebasing all the time. This probably comes down to personal style. If you like the commits to be "pre-squashed" I will do so next time.

alphaFred committed:
mimxrt/machine_pin: Fixed review comments.

dpgeorge commented:
> I thought it would be better for me to keep track when I push most of my commits and do not perform heavy rebasing all the time

Yes that's definitely a good way to do it, makes re-reviewing a lot easier.

Since this is a large PR it's hard to get the commits right first time.
